### PR TITLE
lakectl log: add option to filter merge commits

### LINKED ISSUE #7441: Give an option to hide merge commits in a branch view in the UI
Similarly to GitHub metrics I guess. We switched our transaction model to a merge of an ephemeral branch, which creates merge commits on success. When going through a branch history, these commits clutter the view, so it would be great to have a filtered view where we can focus on the actual additions.

cc @maxmynter

### PR DESCRIPTION
**Pull request by ItamarYuran**
Closes #7441

Added an option to filter merge commits in the log command. The flag `--no-merges` will filter every commit that has more than one parent and thus also import commits. The implementation concerns only lakectl at the moment. Filtering is done at The server's side similarly to `--first-parent` flag.

### Testing Details

Tested only manually, original test pass (after a slight modification) 

## Additional info

Screenshot 2024-09-08 at 10 36 14

### Contact Details
itamar.yuran@treeverse.io

### CODE REVIEW

File: pkg/graveler/ref/commit_iterator.go [hunk 1]
```diff
@@ -116,41 +119,63 @@ func (ci *CommitIterator) Next() bool {
...
+			// skip commits that are older than since time
+			if ci.since != nil && rec.Commit.CreationDate.Before(*ci.since) {
+				continue
+			}
+
+			heap.Push(&ci.queue, rec)
 		}
 
-		heap.Push(&ci.queue, rec)
+		return true
```
Comment (itaiad200) [inline pkg/graveler/ref/commit_iterator.go:172]: Impressive - One of the more delicate areas in Graveler and it seems really thought through.

For me, as a reviewer, it's near impossible to follow the logic that is applied here without comments walking me through the decisions taken in every branch of this code.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Fearful PR in one of the most key areas of Graveler - well done.

In order for me to effectively review this PR, I'm going to request 2 things:
1. Document the new logic in the commit iterator. It's hard to follow this delicate area logic change without this.
2. Tests. First it seems that some tests fail in our CI (not sure if these are related or not). Also - `commit_iterator` needs some testing for the new functionality (perhaps more testing for existing functionality too), lakectl requires tests for the new flag, probably in esti.

**Comment by N-o-Z** [comment]
@ItamarYuran please add missing labels to PR

File: api/swagger.yml [hunk 2]
```diff
@@ -3540,6 +3540,11 @@ paths:
           description: if set to true, follow only the first parent upon reaching a merge commit
           schema:
             type: boolean
+        - in: query
```
Comment (arielshaqed) [inline api/swagger.yml:3543]: Not a fan of parameters with negative names. They're mostly okay as commandline flag, less so in an api. 

However here I have no gear alternative to suggest: the default has to remain with merges. If you *do* have a good alternative I'm very open to suggestions here. 

Otherwise let's go with this one! If we keep on having to add flags, we'll anyway need to end up with an expression parser / GraphQL. So I'm willing to wait until we need another flag.

More importantly is why we need an API change if we just want to filter; that appears on the main comment for this review.

File: pkg/graveler/ref/commit_iterator.go [hunk 3]
```diff
@@ -115,42 +118,75 @@ func (ci *CommitIterator) Next() bool {
...
-		if _, visited := ci.visit[p]; visited {
+	// running as long as we have commit in queue, popping one each iteration
+	for ci.queue.Len() > 0 {
+		ci.value = heap.Pop(&ci.queue).(*graveler.CommitRecord)
+
+		// case --no-merges flag, and there is more than one parent, treating only the first one
+		if ci.noMerges && len(ci.value.Parents) > 1 {
+			firstParent := ci.value.Parents[0]
+
+			// case parent was visited, continuing.
```
Comment (arielshaqed) [inline pkg/graveler/ref/commit_iterator.go:129]: I don't understand.
1. Here no-merges implies first-parent. Why? I would expect no-merges to do just one thing. 
1. How does this code work when there are 2 merges one after the other? firstParent Will also be a merge and the code should skip it.
1. (Related) Just skip the current record of its a merge and keep going. This is a filter operation, let the next iteration do its thing

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Blocking about possible incorrectness.
Also this code needs tests, I assume they'll be in the next commit on this PR so I'm blocking but you probably already know why. 

With that it is the way, I would not like this code to be in the server, and this is my primary reason to block. A client like lakectl can and should filter on its side. There is no efficiency gain from filtering on the server side, most of the slowness comes from the server needing to iterate over many commits. It is also very easy to perform on the high-level python SDK.
Just iterate over calls to log-commits in the client and filter out commits with >1 parent. It will be simpler code and allow filtering as desired.

**Comment by ItamarYuran** [comment]
Thank You all for your reviews.
Completely changed attitude to a client side implementation.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
First of all, I really like this code!

There is some nastiness (an actual tech debt, even a bug) in the original code, that became bad when we add filtering.  If you specify `--amount` (the case `amount != 0` on l. 177) then the original code will never return >1000 results, even if `amount == 2000`.  The new code compounds this: if the first page does not contain even `amount` results, then it will perform a single call.  So it will display `< amount` results.  For instance, say I specify `--amount=10`.  Then the code _asks for 10 log entries_, and displays only those which are not merges.  So it will _often_ fail to display exactly 10 results.

I would suggest:
1. Change the loop to reduce `amount` by the number of results actually output.
   - Exit the loop if we get to `amount <= 0`.
1. When setting up initial pagination, ask for `amount` results if not filtering, or `3*amount` results if we _are_ filtering.  This is a heuristic: it guesses that at most 2/3 commits will be merges, so it asks for more.  But it _always_ works even if it guessed wrong - it just calls lakeFS more times than it could have.

WDYT?

**Comment by ItamarYuran** [comment]
@arielshaqed As a result of using the heuristic, while using both `--no-merges` & `--amount` user would usually get more results than asked

File: cmd/lakectl/cmd/log.go [hunk 4]
```diff
@@ -66,6 +66,19 @@ func (d *dotWriter) Write(commits []apigen.Commit) {
 	}
 }
 
+// filter merge commits, used for --no-merges flag
+func filterMergeCommits(commits []apigen.Commit) []apigen.Commit {
+	filteredCommits := make([]apigen.Commit, 0, len(commits))
```
Comment (itaiad200) [inline cmd/lakectl/cmd/log.go:71]: I don't think you can really rely on the length of commits, all you know is that `0<= len(filteredCommits)<=len(commits)`. 
It's really an extremely minor performance issue, but I rather not allocate more than I need
```suggestion
	var filteredCommits []apigen.Commit
```

File: cmd/lakectl/cmd/log.go [hunk 5]
```diff
@@ -100,6 +114,10 @@ var logCmd = &cobra.Command{
 		if amountForPagination <= 0 {
 			amountForPagination = internalPageSize
 		}
+		// case --no-merges & --amount, ask for more results since some will filter out
+		if noMerges && amountForPagination < maxAmountNoMerges {
+			amountForPagination *= 3
```
Comment (itaiad200) [inline cmd/lakectl/cmd/log.go:119]: Now I get the `333` - it's directly linked to 3. I wonder how our linter doesn't complain about this magic number.

File: cmd/lakectl/cmd/log.go [hunk 6]
```diff
@@ -151,13 +169,19 @@ var logCmd = &cobra.Command{
 				},
 			}
 
+			// case --no-merges, filter commits and subtract that amount from amount desired
+			if noMerges {
+				data.Commits = filterMergeCommits(data.Commits)
+				amount = amount - (3 * len(data.Commits))
```
Comment (itaiad200) [inline cmd/lakectl/cmd/log.go:175]: Why is it tripled here?

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks! Some very good progress here. Besides my comments, we're missing some tests coverage for the new flag

File: cmd/lakectl/cmd/log.go (hunk 6 above)
Comment (ItamarYuran) [reply cmd/lakectl/cmd/log.go:175]: My thought was that if we use that heuristic, in this way we stay true to it. We asked for three times the amount we needed to get ~1/3 results of that. Updating the amount this way will keep that heuristic in every request.

File: cmd/lakectl/cmd/log.go (hunk 4 above)
Comment (ItamarYuran) [reply cmd/lakectl/cmd/log.go:71]: That was my initial proposal, @arielshaqed suggested that

**Comment by N-o-Z** [comment]
@ItamarYuran Can we refactor the PR title from: "7441 give an option to hide merge commits in a branch in lakectl"
to "lakectl log: add option to filter merge commits"?
More concise and there's no need to reference the linked issue as we add it in the description

File: cmd/lakectl/cmd/log.go (hunk 6 above)
Comment (itaiad200) [reply cmd/lakectl/cmd/log.go:175]: It's better if we have a single location to handle amount heuristics. Can you try to combine this block and the above one (lines 114-120) into one? Preferably where the loop starts

**Comment by ItamarYuran** [comment]
@arielshaqed It works! at least it seems like. Failures occurring because of some other reason

File: cmd/lakectl/cmd/common_helpers.go [hunk 7]
```diff
@@ -41,6 +41,10 @@ const (
 	internalPageSize           = 1000 // when retrieving all records, use this page size under the hood
 	defaultAmountArgumentValue = 100  // when no amount is specified, use this value for the argument
 
+	// when using --no-merges & amount, this const is the upper limit to use huristic.
```
Comment (arielshaqed) [inline cmd/lakectl/cmd/common_helpers.go:44]: Nit:
```suggestion
	// when using --no-merges & amount, this const is the upper limit to use heuristic.
```

Comment (arielshaqed) [inline cmd/lakectl/cmd/common_helpers.go:44]: The comment also says "3".  So that constant needs to appear here, too.

File: esti/lakectl_test.go [hunk 8]
```diff
@@ -338,6 +338,59 @@ func TestLakectlMergeAndStrategies(t *testing.T) {
 	RunCmdAndVerifySuccessWithFile(t, Lakectl()+" fs ls lakefs://"+repoName+"/"+featureBranch+"/", false, "lakectl_fs_ls_1_file", vars)
 }
 
+func TestLakectlLogNoMergesWithCommitsAndMerges(t *testing.T) {
+
```
Comment (arielshaqed) [inline esti/lakectl_test.go:342]: Nit: please remove this blank line, it does not aid readability.

**Review by arielshaqed — COMMENTED** [review]
> ... Failures occurring because of some other reason

Unfortunately all Estis seem to fail at the same spot.  If we pull this our CI will break.  So _it doesn't matter_ if the bad test code was added here or not, I cannot approve.

What can we do?

First, skip your added test (add `t.Skip()` at the beginning) and get Esti to run on it again.  Does it still fail?

If it _stopped_ failing, we need to look at the added test - or perhaps t.Skip() _another_ test to see if it's just one extra test which breaks everything.  If it _still_ fails, we will need to understand why (whether) the trunk fails, and open an issue to fix trunk.

Sorry.

**Comment by arielshaqed** [comment]
Note that [these runs on master](https://github.com/treeverse/lakeFS/actions/workflows/esti.yaml?query=branch%3Amaster) all seem to succeed; I started another one right now, but I would expect it also to succeed.

**Comment by arielshaqed** [comment]
I also re-ran Esti on this branch.  This will give us more confidence whether the problem is on trunk or only on this branch.

**Comment by arielshaqed** [comment]
I think [this](https://github.com/treeverse/lakeFS/actions/runs/11093694546/job/30820345219?pr=8142) means that Esti actually discovered an issue with the code.  It is getting a wrong result when logging after a merge. 🙀

**Review by arielshaqed — APPROVED** [review]
Thanks, great feature!
